### ISSUE TITLE
Use wildcard SSH host names

### ISSUE THREAD

**Issue by code-asher** [coder/coder#14986]
I would like to propose that we configure SSH with a wildcard instead of individual host names.  For example:

```
coder.*
  ProxyCommand coder %h
```

And in `ssh.go` we trim out the `coder.` prefix.

Mainly, this is to support the IDEs.  VS Code already does this with a special `vscodessh` sub-command but I think we should roll that up into `ssh`, then JetBrains can also benefit.

The main reason is that the IDEs support connecting to anyone's workspace (for admins).  That means potentially hundreds of SSH config entries, and potentially hundreds of calls to template resources, since that has to be [done for every workspace](https://github.com/coder/coder/blob/ea3b13c78e2bcc15317ba74070084b61233b7589/cli/configssh.go#L161), at least the ones that are off, to get the agent names.

It could also eventually remove the need for the Gateway and VS Code plugins to implement their own versions of `config-ssh` as they currently do, but this would need one more tweak to include the deployment in the host (for example `coder-vscode.dev.coder.com--*` is the host we use in VS Code).

It would eliminate the resource fetch loop in `config-ssh` as well.

Alternatives:
- A separate `idessh` sub-command that does this, maybe just rename `vscodessh` and add a deprecated alias.
- Gate it behind flags (`config-ssh --wildcard` and `ssh --wildcard` or some such).
- Skip changing `config-ssh` for now, and add `ssh --strip-prefix=coder-vscode.dev.coder.com--` or some such flag to at least allow the IDEs to configure their own wildcard host.
- From the IDEs, we could instead do some inline shell magic to strip out the prefix, but I am not sure what Windows support looks like.

```
coder-vscode.dev.coder.com--*
  ProxyCommand coder $(echo "%h" | sed 's/coder-vscode.dev.coder.com--//')
```

**Comment by code-asher** [comment]
Another alternative that at least partially works around the slowness, in JetBrains we could only add hosts to the config for the currently connected workspace and any in the recent connections.  A wildcard would be nice but maybe this could be a temporary bandaid.

**Comment by code-asher** [comment]
Or, if we actually parse the host names, we can more smartly update the config and only make requests for workspaces that are new.  Currently we interact with the config as one big block, so it is all or nothing.

**Comment by code-asher** [comment]
Another alternative, if the main workspaces query included all the agents even for off workspaces, this would at least make it faster probably.

**Comment by ammario** [comment]
I enjoy this idea and think we should strive for one opinionated method for SSH configuration. The one drawback to wildcards (which might be a worthwhile cost) is users lose autocomplete when typing `ssh coder.<TAB>`.

**Comment by code-asher** [comment]
Great point, I had not considered that.  Hmmm maybe we can provide completions to compensate (`coder ssh --completions` or something, idk if there is some standard flag folks tend to use, but then this would be placed in `/etc/bash_completions` or wherever).  🤔

**Comment by ammario** [comment]
Ideally we would keep all completions coming in through `coder completion`.

**Comment by code-asher** [comment]
Oof 🤦 I missed that.  Yup, makes sense.

**Comment by Emyrk** [comment]
Is there a way to do dynamic completions? Like can the autocomplete do a `coder ls` and pull from that list?

**Comment by ethanndickson** [comment]
> Is there a way to do dynamic completions? Like can the autocomplete do a coder ls and pull from that list?

Yep, you can override the completion handler in serpent to do it all without touching any shell :)

**Comment by ethanndickson** [comment]
Not override, sorry, it used to override - now the custom handler gets run in addition.

**Comment by code-asher** [comment]
Ah yeah if I understand correctly, that can be used to add completions for itself (the `coder` binary) but it would not be able to add completions for some other random external binary (like `ssh`).

Basically we want it so when you type `ssh <TAB>` in the terminal it will auto-complete with Coder workspaces.

**Comment by code-asher** [comment]
When this gets assigned, I will chat with whoever gets it to make sure my thoughts came across correctly.

**Comment by ethanndickson** [comment]
ahh yeah I see, my bad!

**Comment by matifali** [comment]
@spikecurtis will this functionality change how Coder Desktop works?

**Comment by spikecurtis** [comment]
> [@spikecurtis](https://github.com/spikecurtis) will this functionality change how Coder Desktop works?

With Coder Desktop, it won't be necessary to use the `coder` CLI binary to create the tunnel to your workspace. 

However, CoderVPN uses `.coder` as a _suffix_ to host names, not a prefix like our SSH config. So, we should start the work to migrate our ssh configuration to follow this pattern, since it tracks the semantic meaning of DNS order, and should avoid name collisions.

**Comment by matifali** [comment]
@bcpeinhardt, Can you help me decide if it's not already done? Thanks.

**Comment by bcpeinhardt** [comment]
@matifali I would say the title of the issue (use wildcard SSH host names) is complete in 2.19 of Coder, but there's other work present in the comments worth looking at (mainly Spikes point about our current configuration not really following the semantics of DNS order). I wonder if we already have an issue for that 🤔 
cc @code-asher

**Comment by matifali** [comment]
Thanks @bcpeinhardt. Yes, for following the semantics of DNS, we now have #16828.
So, I am closing this as completed.

Also, tagging @ThomasK33 to have more context on why we are switching to a suffix.

**Comment by octoxan** [comment]
This broke so many things for us. This should have been optional.

**Comment by spikecurtis** [comment]
@octoxan Can you tell us any more information about what broke and how?
